rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,330 @@
+
+The host app will follow the default MVC structure of Rails applications, with some additional restrictions on what goes where that are frontend specific (listed below)
+
+```
+├── app
+│   ├── assets
+│   ├── controllers
+│   ├── helpers
+│   ├── models
+│   ├── presenters
+│   ├── services
+│   ├── views
+├── config
+│   ├── data
+│   ├── locales
+├── lib
+│   ├── tasks
+└── .gitignore
+```
+
+Under `/app`:
+
+`models`
+- Frontend apps do not have database records, so models here are created to store data received from api calls (eg content items from content store), process data that will be written to apis (eg feedback items to zendesk), and information ingested from YAML/JSON/CSVs.
+
+`services`
+- Api calls for search and content-store, and apis external to GOV.UK
+
+`presenters`
+- Presentation related transformations on model data. It is a wrapper around a model for use in a view.
+
+`helpers`
+- General utility methods for a view.
+
+Under `/lib`
+
+`tasks`
+- Rake tasks
+
+Use `lib` rather than `app/lib` for custom code that enhances the app but doesn’t fit into a directory in `app`
+
+Under `/config`
+
+`data`
+- YAML, JSON, CSV config files. Data files should be placed here rather than under `/lib/data`.
+
+`locales`
+- Static content like lists of contact information should ideally be stored in the locale files to promote localisation.
+
+Namespaces will be used to group functionality e.g. for StepBySteps, and resources in the routes file. This structure will be documented in README of the host app.
+Each application will be analysed before being consolidated to ensure redundant code is not being moved over.
+
+### How the proposal will be carried out
+This is obviously a very large piece of work that will take some time to complete. We have identified the following steps that will be needed.
+
+* Prepare `frontend` to be the host
+    * convert frontend tests to RSpec
+    * modernise the tests
+    * audit `frontend` and make changes to the frontend directory structure to match “new layout”
+    * remove redundant code from frontend (non-blocking)
+    * remove publishing_api tasks from frontend (non-blocking)
+* Merge `feedback` into `frontend`
+    * use as a test case and use that to inform our next steps
+* Write a roadmap for app consolidation
+    * will be handled separately
+    * might be simply the order in which applications will be merged into the host
+    * or could be some apps merged together before merging into the host
+    * or some routes merging ahead of others
+* Consider and work on expected issues (see below)
+* Work to merge remaining apps
+
+### Steps as part of a merge
+We anticipate that any merge will likely involve some or all of the following steps.
+
+* Identify any issues with existing code
+    * size these issues
+    * either fix issues or record them as tech debt - we need to avoid trying to fix everything and concentrate on progressing consolidation
```

Comment:
Candidate for should be fixed before moving over https://github.com/alphagov/government-frontend/tree/main/app/views/histories

Reply:
Ah, yes this content should be moved into a locale file. I think that can be "fixed" as part of the migration.

Reply:
Do we know the history of why the history pages aren't just normal content items? Is there a way we can remove them entirely from government-frontend and turn them into editable items?

Reply:
Oh they're published by [government-frontend](https://github.com/alphagov/government-frontend/blob/main/lib/publish_static_pages.rb) at the moment 😬 

It is a bit of a tangled mess what came first? [the template or the content item](https://github.com/alphagov/government-frontend/blob/main/lib/publish_static_pages.rb#L92)? 😬 

As a first step the task that publishes the pages should be moved to a publishing app. They at least do for the most part seem to follow a pattern: breadcrumbs, contents list, image, govspeak.